### PR TITLE
Fix ELU gradient precision for negative inputs

### LINKED ISSUE #124833: tf.nn.elu returns zero gradient for negative float64 input where analytic derivative is finite
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0

### Custom code

Yes

### OS platform and distribution

 Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.nn.elu` returns an incorrect first derivative for a negative float64 input.

For ELU with default `alpha=1`, the negative branch is `exp(x) - 1`, so the derivative for `x < 0` should be `exp(x)`. At `x = -40.0`, the analytic high-precision reference is approximately `4.248354255291589e-18`, which is a finite normal float64 value.

However, TensorFlow autodiff returns `0.0` for the gradient at this input. PyTorch computes the expected finite value for the same derivative, so this appears to be a backward-pass issue rather than a float64 limitation.

Expected behavior?

The gradient of `tf.nn.elu` at `x = -40.0` should be close to `exp(-40.0)`, approximately `4.248354255291589e-18`, instead of `0.0`.

### Standalone code to reproduce the issue

```shell
import os
os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

import math
import tensorflow as tf

x = -40.0
truth = math.exp(x)
rtol = 1e-6

v = tf.Variable(x, dtype=tf.float64)

with tf.GradientTape() as tape:
    y = tf.nn.elu(v)

g = tape.gradient(y, v)

actual = float(g)
rel_err = abs(actual - truth) / abs(truth)

print(f"x: {x!r}")
print(f"tensorflow gradient: {actual!r}")
print(f"expected gradient: {truth!r}")
print(f"relative error: {rel_err:.3e}")

if rel_err > rtol:
    print("BUG REPRODUCED: tf.nn.elu gradient is wrong for negative float64 input")
else:
    print("not reproduced")
```

### Relevant log output

```shell
x: -40.0
tensorflow gradient: 0.0
expected gradient: 4.248354255291589e-18
relative error: 1.000e+00
BUG REPRODUCED: tf.nn.elu gradient is wrong for negative float64 input
```

### PR DESCRIPTION
**Pull request by GodlyDonuts**
Fixes #124833.

## What changed

- Compute the ELU derivative from the input as `exp(min(x, 0.0))` instead of reconstructing it as `output + 1`. The output rounds to exactly `-1` for sufficiently negative inputs even when the derivative is still representable.
- Update the generated GradientTape exclusion metadata so eager execution retains the ELU input and can release the now-unused output.
- Add `float32` and `float64` regression coverage in both graph and eager modes.

## Validation

- Confirmed the expanded regression fails with the previous TensorFlow 2.20 implementation for both dtypes in graph and eager modes (`0.0` instead of the finite analytic derivatives).
- Verified the stable derivative produces `2.0611537e-09` for `float32` at `-20` and `4.248354255291589e-18` for `float64` at `-40`.
- `python3 -m py_compile tensorflow/python/ops/nn_grad.py tensorflow/python/ops/nn_grad_test.py`
- TensorFlow PyLint configuration: 10.00/10 (excluding two unrelated existing inference false positives in `nn_grad.py`).
- `git diff --check`

### CODE REVIEW

File: tensorflow/python/ops/nn_grad.py [hunk 1]
```diff
@@ -458,7 +458,10 @@ def _LeakyReluGradGrad(op: ops.Operation, grad):
 
 @ops.RegisterGradient("Elu")
 def _EluGrad(op: ops.Operation, grad):
-  return gen_nn_ops.elu_grad(grad, op.outputs[0])
+  x = op.inputs[0]
+  # Computing the gradient from the ELU output loses precision when exp(x) is
+  # too small to affect -1. Compute it directly from the input instead.
+  return grad * math_ops.exp(math_ops.minimum(x, 0))
```
[the remark this answers is not in the document]
Comment (GodlyDonuts) [reply tensorflow/python/ops/nn_grad.py]: Addressed in d9d7231de14: the minimum bound is now explicitly `0.0`. I will keep the related SELU issue separate so this PR remains scoped to #124833.

File: tensorflow/python/ops/nn_grad_test.py [hunk 2]
```diff
@@ -181,6 +181,17 @@ def testDepthwiseConv2dBackpropFilterGrad(self):
...
+  @test_util.run_in_graph_and_eager_modes
+  def testEluGradPreservesSmallNegativeValues(self):
+    inputs = constant_op.constant(-40.0, dtype=dtypes.float64)
+    with backprop.GradientTape() as tape:
+      tape.watch(inputs)
+      elu = gen_nn_ops.elu(inputs)
+
+    elu_grad = tape.gradient(elu, inputs)
+    self.assertAllClose(
+        np.exp(-40.0), self.evaluate(elu_grad), rtol=1e-14, atol=0)
```
[the remark this answers is not in the document]
Comment (GodlyDonuts) [reply tensorflow/python/ops/nn_grad_test.py]: Addressed in d9d7231de14: the regression now covers both float32 at -20 and float64 at -40 in graph and eager modes, with dtype-appropriate tolerances.

**Review by dmiltr3 — APPROVED** [review]
